File: ruby/tests/list_ops.rb
require "json"

class ListOpsTest
  attr_reader :name

  def initialize(ops_per_iter = 1_000_000)
    @name = "ListOps"
    @ops_per_iter = ops_per_iter
  end

  def workload
    arr = []
    # Append phase
    5000.times { |i| arr << i }
    # Random insertions
    500.times do
      idx = rand(0..arr.length)
      arr.insert(idx, rand(0..100))
    end
    # Random deletions
    500.times do
      break if arr.empty?
      idx = rand(0...arr.length)
      arr.delete_at(idx)
    end
    # Access phase
    s = 0
    1000.times { s += arr[rand(0...arr.length)] }
    s
  end

  def run_once(iterations)
    reps = [1, iterations / 1000].max
    t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    total = 0
    reps.times { total += workload }
    t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    duration = t1 - t0
    ops_done = reps * 1000
    {
      "time_s" => duration,
      "ops" => ops_done,
      "ops_per_sec" => ops_done / (duration > 0 ? duration : 1e-9),
      "acc" => total,
      "gc_stat_after" => GC.stat
    }
  end

  def run(runs = 5, iterations = nil)
    iterations ||= @ops_per_iter
    results = []
    runs.times do
      results << run_once(iterations)
      GC.start
    end
    times = results.map { |r| r["time_s"] }.sort
    ops = results.map { |r| r["ops_per_sec"] }.sort
    {
      "name" => @name,
      "runs" => runs,
      "median_time_s" => times[times.length / 2],
      "median_ops_per_sec" => ops[ops.length / 2],
      "raw" => results
    }
  end
end

if __FILE__ == $0
  t = ListOpsTest.new
  puts JSON.pretty_generate(t.run)
end

File: python/tests/list_ops.py
import time
import random
import json
import gc
import os

class ListOpsTest:
    def __init__(self, ops_per_iter=1_000_000, name="ListOps"):
        self.name = name
        self.ops_per_iter = ops_per_iter

    def workload(self):
        lst = []

        # Append phase
        for i in range(5000):
            lst.append(i)

        # Random insertions
        for _ in range(500):
            idx = random.randint(0, len(lst))
            lst.insert(idx, random.randint(0, 100))

        # Random deletions
        for _ in range(500):
            if lst:
                idx = random.randint(0, len(lst) - 1)
                del lst[idx]

        # Access phase
        s = 0
        for _ in range(1000):
            s += lst[random.randint(0, len(lst) - 1)]

        return s

    def run_once(self, iterations):
        reps = max(1, iterations // 1000, 50)
        t0 = time.perf_counter()
        total = 0

        for _ in range(reps):
            total += self.workload()

        t1 = time.perf_counter()
        duration = t1 - t0
        ops_done = reps * 1000

        return {
            "time_s": duration,
            "ops": ops_done,
            "ops_per_sec": ops_done / (duration if duration > 0 else 1e-9),
            "acc": total,
            "gc_stat_after": gc.get_stats()[-1] if hasattr(gc, "get_stats") else None
        }

    def run(self, runs=None, iterations=None):
        iterations = iterations or self.ops_per_iter
        runs = runs or 5
        results = []

        for _ in range(runs):
            results.append(self.run_once(iterations))
            gc.collect()

        times = sorted(r["time_s"] for r in results)
        ops = sorted(r["ops_per_sec"] for r in results)
        median_time = times[len(times)//2]
        median_ops = ops[len(ops)//2]

        return {
            "name": self.name,
            "runs": runs,
            "median_time_s": median_time,
            "median_ops_per_sec": median_ops,
            "raw": results
        }

if __name__ == "__main__":
    test = ListOpsTest()
    result = test.run()
    with open("results_list_ops_python.json", "w") as f:
        json.dump(result, f, indent=2)
    print(f"{result['name']}: Median Time={result['median_time_s']:.4f}s | "
          f"Ops/sec={result['median_ops_per_sec']:.2f}")
